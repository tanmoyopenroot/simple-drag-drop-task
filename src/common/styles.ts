import styledComponents from 'styled-components';

export interface IGroupedRowProps {
  justifyContent?: string;
}

export const GroupedRow = styledComponents.div<IGroupedRowProps>`
  display: flex;
  flex-direction: row;
  justify-content: ${props => props.justifyContent || 'space-between'};
  align-items: center;
`;

export interface IGroupedColumnProps {
  justifyContent?: string;
}

export const GroupedColumn = styledComponents.div<IGroupedColumnProps>`
  display: flex;
  flex-direction: column;
  justify-content: ${props => props.justifyContent || 'space-between'};
  align-items: center;
`;
