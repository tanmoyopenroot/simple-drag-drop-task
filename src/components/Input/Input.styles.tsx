import styledComponents from 'styled-components';

export const InputElement = styledComponents.textarea`
  color: ${props => props.theme.PRIMARY_TEXT};
  font-size: ${props => props.theme.FONT.MEDIUM};
  width: 95%;
  border: 0px;
  outline: 0px;
  padding: 0px;
  background: transparent;
`;
