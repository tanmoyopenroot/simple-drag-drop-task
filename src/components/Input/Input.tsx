import * as React from 'react';

import { InputElement } from './Input.styles';

enum InputType {
  TEXT = 'text',
}

enum KEYS {
  ENTER = 'Enter',
}

export interface IInputProps {
  autoFocus?: boolean;
  defaultValue?: string;
  type?: InputType;
  placeholder?: string;
  onChange?: (value: string) => void;
  onBlur?: (value: string) => void;
}

export interface IInputState {
  query: string;
}

class Input extends React.PureComponent<IInputProps, IInputState> {
  public static Type = InputType;
  public static defaultProps: Partial<IInputProps> = {
    autoFocus: false,
    defaultValue: '',
    placeholder: '',
    type: InputType.TEXT,
  };

  public readonly state: IInputState = {
    query: '',
  };

  public componentDidMount() {
    const { defaultValue } = this.props;

    if (defaultValue) {
      this.setState({ query: defaultValue });
    }
  }

  private handleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const { query } = this.state;
    const {
      onChange,
      onBlur,
    } = this.props;

    if (e.key === KEYS.ENTER && onChange) {
      onChange(query);
    }

    if (e.key === KEYS.ENTER && onBlur) {
      onBlur(query);
    }
  }

  private handleChange = (e: React.FormEvent<HTMLElement>) => {
    const query = (e.target as HTMLInputElement).value;
    const { onChange } = this.props;

    this.setState({ query });
    if (onChange) {
      onChange(query);
    }
  }

  private handleBlur = (e: React.FormEvent<HTMLElement>) => {
    const query = (e.target as HTMLInputElement).value;
    const { onBlur } = this.props;

    this.setState({ query });
    if (onBlur) {
      onBlur(query);
    }
  }

  render() {
    const { query } = this.state;
    const {
      autoFocus,
      type,
      placeholder,
    } = this.props;

    return(
      <InputElement
        autoFocus={autoFocus}
        type={type}
        value={query}
        placeholder={placeholder}
        onKeyDown={this.handleKeyDown}
        onChange={this.handleChange}
        onBlur={this.handleBlur}
      />
    );
  }
}

export default Input;
