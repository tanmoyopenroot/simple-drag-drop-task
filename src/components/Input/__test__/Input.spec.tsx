import * as React from 'react';
import { mount } from 'enzyme';

import Input, { IInputProps, IInputState } from '../Input';
import { InputElement } from '../Input.styles';

describe('Testing <Input />', () => {
  const props: IInputProps = {
    autoFocus: true,
    defaultValue: 'abc',
    type: Input.Type.TEXT,
    onBlur: jest.fn(),
    onChange: jest.fn(),
  };
  const wrapper = mount<IInputProps, IInputState>(<Input {...props}/>);

  it('should be able to render with props', () => {
    expect(wrapper.length).toEqual(1);
    expect(wrapper.children().length).toBe(1);
    expect(wrapper.props().autoFocus).toEqual(props.autoFocus);
    expect(wrapper.props().defaultValue).toEqual(props.defaultValue);
    expect(wrapper.props().type).toEqual(props.type);
    expect(wrapper.props().onBlur).toEqual(props.onBlur);
    expect(wrapper.props().onChange).toEqual(props.onChange);
    expect(wrapper).toMatchSnapshot();
  });

  it('should be able to change input value', () => {
    expect(wrapper.state().query).toBe(props.defaultValue);

    wrapper
      .find(InputElement)
      .simulate('change', { target: { value: 'newChange' } });
    expect(wrapper.state().query).toBe('newChange');
    expect(props.onChange).toBeCalledWith('newChange');

    wrapper
      .find(InputElement)
      .simulate('blur', { target: { value: 'newBlur' } });
    expect(wrapper.state().query).toBe('newBlur');
    expect(props.onBlur).toBeCalledWith('newBlur');
  });
});
