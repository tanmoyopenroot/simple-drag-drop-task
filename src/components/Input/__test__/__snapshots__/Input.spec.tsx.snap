// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <Input /> should be able to render with props 1`] = `ReactWrapper {}`;
